Create React App を使った SPA のメモアプリを作成した

nakamu-kazu222 opened the pull request:
create-vue / create-react-app を使ってSPAとして以下が実装しているメモアプリを作る
## 実装内容
- 一覧
  - メモの1行目をタイトルとして一覧表示する
  - タイトルをクリックするとそのメモの編集状態に移行する
- 詳細
  - 編集状態＝詳細
- 追加
  - ＋をクリックすると「新規メモ」というメモファイルが作成され、編集状態に移行する
- 編集
  - テキストエリアにメモの内容を表示し、編集できる
  - 編集ボタンをクリックすると保存される
- 削除
  - 編集状態で削除ボタンをクリックするとメモは削除される
  - そのあとは一覧の状態に移行する
- 保存先ストレージ
  - LocalStorageやFirebaseなど、永続化できれば自由

## 条件
以下の条件を満たすように作成する
- React コンポーネントを 2 つ以上作成する
- JSX を利用する
- useStateを利用する

nakamu-kazu222 committed:
READMEにメモアプリの概要を記載した

nakamu-kazu222 committed:
READMEを更新した

nakamu-kazu222 committed:
メモ一覧を表示するコンポーネントを作成した

nakamu-kazu222 committed:
メモを記載するコンポーネントを作成した

nakamu-kazu222 committed:
メモデータを格納するコンポーネントを作成した

nakamu-kazu222 committed:
READMEを更新した

実行方法を追加した

nakamu-kazu222 committed:
onAddを使用して新しいメモを作成できるようにした

nakamu-kazu222 committed:
すべての機能を実装した

また、CSSでレイアウトを整えた

nakamu-kazu222 committed:
ESLint と Prettier の実行

nakamu-kazu222 committed:
不要なファイルを削除した

nakamu-kazu222 committed:
ESLint と Prettier の環境設定をした

nakamu-kazu222 committed:
CSSの調整をした

nakamu-kazu222 committed:
npx prettier --write 実行

nakamu-kazu222 committed:
削除した後の処理は必要ないため削除した

yuuu commented on my-memo-app/src/App.js:
```diff
@@ -1,25 +1,82 @@
+  const [memos, setMemos] = useState(
+    JSON.parse(localStorage.getItem("memos")) || [],
+  );
+  const [idCounter, setIdCounter] = useState(1);
```
カウンターを `state` にしているのでこのコンポーネントがリレンダリングされるたび(ブラウザのリロードでも)にカウント結果が1に戻ってしまうのではないでしょうか？ 🤔

nakamu-kazu222 replied:
すみません、私の理解ができていないと思いますが、
`setIdCounter(idCounter + 1)` によって idCounter が増加し、新しいメモに一意のIDが割り当てられると思い設定しました

yuuu replied:
リロードをしなければその理解で問題ないですが、リロードをするとidCounterは1に戻っています。そうするとリロード前に作成したメモとリロード後に作成したメモはIDが重複することになるので、対策を検討いただきたいです。

nakamu-kazu222 replied:
すみません、理解しました

localStorageを使用してidCounterの値を保存し、アプリケーションがリロードされてもその値を保持できるようにしました

yuuu commented on my-memo-app/src/App.js:
```diff
@@ -1,25 +1,82 @@
+  };
+
+  const handleEditMemo = (editId, newText) => {
+    const updatedMemos = memos.map((memo) => {
```
ここは `find` を使うようにするともう少しコードがスッキリしますね。

nakamu-kazu222 replied:
`find`を使用してidを選択するようにしました

yuuu commented on my-memo-app/src/components/WriteMemo.js:
```diff
@@ -0,0 +1,54 @@
+import React, { useState, useEffect } from "react";
+
+const WriteMemo = ({ selectedMemo, onAdd, onEdit, onRemove }) => {
+  const [text, setText] = useState("");
```
ここを以下のように書けば `useEffect` が消せそうですね。

```javascript
const [text, setText] = useState(selectedMemo&.text || "");
```

nakamu-kazu222 replied:
アドバイスいただいたコードを使用して、`useEffect`を使わずに初期値を設定できるようにしました

yuuu commented on my-memo-app/src/App.js:
```diff
@@ -1,25 +1,82 @@
+          </button>
+        </div>
+        <div>
+          {showWriteMemo && (
```
`showWriteMemo` と `selectedMemo` は常にセットで更新しているので `selectedMemo` にまとめてはどうでしょう？
そうすれば `WriteMemo` 内で行っている `selectedMemo` の存在チェックもいらなくなるはずです。

nakamu-kazu222 replied:
`showWriteMemo`を削除して、`selectedMemo`がtrueの時に`WriteMemo`を表示するようにしました

nakamu-kazu222 committed:
findを使用してidを選択するようにした

nakamu-kazu222 committed:
useEffectを使わずに初期値を設定できるようにした

nakamu-kazu222 committed:
showWriteMemoを削除して、selectedMemoがtrueの時にWriteMemoを表示するようにした

showWriteMemo と selectedMemo は常にセットで更新しており、
selectedMemoがtrueであれば、showWriteMemoがtrueになっているので変更した

nakamu-kazu222 committed:
id順に表示されるようにした

yuuu commented on my-memo-app/src/App.js:
```diff
@@ -28,16 +28,19 @@ const App = () => {
+    if (updatedMemos) {
+      const updatedMemo = { ...updatedMemos, text: newText };
+      const updatedMemosArray = [
+        ...memos.filter((memo) => memo.id !== editId),
```
updateMemosを更新して、memosをセットするだけで良いと思うのですが、このような処理をしているのはなぜでしょうか？この方法だと、配列の順番が変わってしまい、そのためにソートをしなければならなくなっている認識です。

nakamu-kazu222 replied:
`map`を使用しないで、`find`を使おうとしまして、上記のような処理になってしまいました

`find`と`map`を使用して修正しました

yuuu replied:
`find` をした後に改めて `map` をする必要がないと感じています。
`find` をして取得できる結果が、その配列内の参照であるためです。以下にサンプルプログラムを記述します。

```javascript
const names = [{ name: 'Taro' }, { name: 'Jiro' }]
const target = names.find(({name}) => name === 'Taro')
target.name = 'Saburo'
console.log(names)
// [{ name: 'Saburo' }, { name: 'Jiro' }]
```

nakamu-kazu222 replied:
理解しました
`find `で取得したメモオブジェクトを`editedMemo.text = newText;`で変更するようにしました
これを`setMemos([...memos])`で更新するようにしました

yuuu replied:
こちら、`find` を使った方法だと一時的とはいえ `state` を破壊してしまうので良くないことに気づきました。修正いただいたあとで大変申し訳ないのですが元の `map` を使った処理に戻して頂いても良いでしょうか？ 🙇

nakamu-kazu222 replied:
`editedMemo.text = newText;`の部分のように直接的なプロパティの変更が良くないということでしょうか？

`map`を使用して更新するようにしました

yuuu replied:
はい、その認識で合っています 🙏 お手数おかけいたしました 🙇

nakamu-kazu222 committed:
localStorageを使用してidCounterの値を保存し、アプリケーションがリロードされてもその値を保持できるようにした

nakamu-kazu222 committed:
findとmapを使用してリファクタリングした

nakamu-kazu222 committed:
find で取得したメモオブジェクトをeditedMemo.text = newText;で変更した

nakamu-kazu222 committed:
find を使った方法では、state を破壊してしまうので変更した
